cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py

```diff
@@ -0,0 +1,317 @@
+        # to the |00...0> state to obtain the MPS state
+        for layer_index in range(max_num_layers):
+            # Compress the MPS from the previous layer to a maximum bond dimension of 2,
+            # |ψ_k> -> |ψ'_k>
+            compressed_mps = disentangled_mps.copy(deep=True)
+
+            # Normalization improves fidelity of the encoding
+            compressed_mps.normalize()
+            compressed_mps.compress(form="left", max_bond=2)
+
+            unitary_layer = self.generate_layer(compressed_mps)
+            unitary_layers.append(unitary_layer)
+
+            # To update the MPS definition, apply the inverse of U_k to disentangle |ψ_k>,
+            # |ψ_(k+1)> = inv(U_k) @ |ψ_k>
+            for i, _ in enumerate(unitary_layer):
+                inverse = unitary_layer[-(i + 1)][1].conj().T
+
+                if inverse.shape[0] == 4:
+                    disentangled_mps.gate_split_(inverse, (i - 1, i))
+                else:
+                    disentangled_mps.gate_(inverse, (i), contract=True)
+
+            # Compress the disentangled MPS to a maximum bond dimension of chi_max
+            # This is to ensure that the disentangled MPS does not grow too large
+            # and improves the fidelity of the encoding
+            disentangled_mps: qtn.MatrixProductState = (  # type: ignore
+                qtn.tensor_1d_compress.tensor_network_1d_compress(
+                    disentangled_mps, max_bond=chi_max
+                )
+            )
+
+            fidelity = np.abs(np.vdot(disentangled_mps.to_dense(), zero_state)) ** 2
+            fidelity = float(np.real(fidelity))
+
+            if fidelity >= self.max_fidelity_threshold:
+                logger.info(f"Reached target fidelity {fidelity}. {layer_index + 1} layers used.")
+                break
+
+        if layer_index == max_num_layers - 1:
+            logger.info(
+                f"Reached fidelity {fidelity} with maximum number of layers {max_num_layers}."
+            )
+
+        # The layers disentangle the MPS to a state close to |00...0>
+        # inv(U_k) ... inv(U_1) |ψ> = |00...0>
+        # So, we have to reverse the layers and apply them to the |00...0> state
+        # to obtain the MPS state
+        # |ψ> = U_1 ... U_k |00...0>
+        unitary_layers.reverse()
+
+        for unitary_layer in unitary_layers:
+            for qubits, unitary in unitary_layer:
+                qubits = list(reversed([mps.L - 1 - q for q in qubits]))
+
+                # In certain cases, floating point errors can cause `unitary`
+                # to not be unitary
+                # We handle this by using SVD to approximate it again with the
+                # closest unitary
+                U, _, Vh = np.linalg.svd(unitary)
+                unitary = U @ Vh
+
+                circuit.append(cirq.ops.MatrixGate(unitary)(*[qr[q] for q in qubits]))
+
+        return circuit
+
+    def __call__(
```

Comment:
Can we convert this to a regular `mps_circuit_from_statevector(statevector, max_num_layers=10, max_fidelity_threshold=0.95)` function and make the `Sequential` class module-internal only?

If so, only the mps_circuit_from_statevector should be exposed in the new `__init__.py` here.